libass/ass_cache.c

```diff
@@ -508,26 +568,41 @@ void ass_cache_empty(Cache *cache)
         while (item) {
             assert(item->size);
             CacheItem *next = item->next;
-            if (item->queue_prev)
-                item->ref_count--;
-            if (item->ref_count)
-                item->cache = NULL;
-            else
+
+            if (!item->queue_prev)
```

Comment:
`ass_cache_empty` is called only from `ass_cache_done` (at which point no remaining external references to the cache can be held), `ass_set_fonts` (which cannot be called concurrently with other functions on the same renderer), and `ass_reconfigure` (which is called only from other public functions that cannot be called concurrently with e.g. rendering).

Reply:
I think there was a plan to make image refcounting public. In that case there can be alive references to cache items. I feel it would be better to make `ass_cache_empty()` safe to concurrent calls of `ass_cache_dec_ref()` in this PR to not forget it later.

Reply:
It'd still be fine as long as we didn't allow calling unref and other functions at the same time. Otherwise, we'd need to generally support concurrent deletion and insertion, which I haven't verified this works for.

Reply:
It would be problematic if `ass_cache_dec_ref(item)` gets called immediately after `while (item) {` for example.

Reply:
Yes, but that could only happen if `ass_cache_dec_ref` and `ass_render_frame` (or the like) were allowed to be called concurrently.